### PR TITLE
Feature 245

### LINKED ISSUE #245: Show error on pictogram upload failed
**Is your feature request related to a problem? Please describe.**
Currently when a Pictogram is uploaded, no feedback is given if an error ocours.

**Describe the solution you'd like**
Something like a pop-up or modal window saying success or giving an error message.

### PR DESCRIPTION
**Pull request by MrPekar98**
fixes #245

### CODE REVIEW

**Comment by Laexander** [comment]
I found two problems while testing on an emulator. When I create a new activity and try to add from the gallery, it immediately gives me a message that it succeeded or failed, before taking me to a screen to pick what picture i want for the pictogram. 
And if i use the back button on the bottom left, to return to the "tilføj fra galleri" screen, instead of choosing a picture, all subsequent messages are failed to upload messages regardless of whether the image was updated, until I return to the weekplanner screen and create a new activity again.

I press the "Tryk for at vælge billede" button and can briefly see a message popup before i go to a screen where i can select an image.
image
Can see the message in the background while it is opening the next screen. 
image



The screen for selecting images. By pressing the back button on the bottom left i go back to the previous screen with the message now displayed, even if I didn't pick a picture. 
image
image

The first time i go through this process after trying to create a new activity it says it was successful.
However if I used the back button without selecting a picture on a previous attempt I get a Upload failed message, even though the upload seemingly succeeded. 
image

Any subsequent attempts, result in a failed message. If I go back to the weekplanner screen and create a new activity again but always choose a picture (and don't use the back button), I always get a success message
image

**Review by Laexander — CHANGES_REQUESTED** [review]
The problems mentioned in the previous comment still exists. The messages (now only the error message) are still displayed when you press the "Tryk for at vælge billede", and using the back button still locks the system into showing error messages until you return to the weekplanner screen.

**Review by Laexander — DISMISSED** [review]

File: lib/screens/upload_image_from_phone_screen.dart [hunk 1]
```diff
@@ -125,19 +126,54 @@ class UploadImageFromPhone extends StatelessWidget {
...
+            StreamBuilder<bool>(
+              stream: _uploadFromGallery.uploadSuccess,
+              builder: (BuildContext context,
+                  AsyncSnapshot<bool> uploadSuccessSnapshot) {
+                return Container(
+                  child: FlatButton(
+                      onPressed: () {
+                        _uploadFromGallery.chooseImageFromGallery();
+
+                        if (uploadSuccessSnapshot.hasData &&
```
Comment (giano574) [inline lib/screens/upload_image_from_phone_screen.dart]: The code for checking upload success is in the wrong place

**Comment by giano574** [comment]
- I get `NoSuchMethodError: The getter 'isNotEmpty' was called on null.` at `package:weekplanner/blocs/upload_from_gallery_bloc.dart:65` after picking an image (which sometimes crashes the app).
- Pressing the back button while on the "Tilføj from galleri" screen goes back to "Ny ugeplan" screen instead of "Piktogram" screen.
- There are no tests to show that the error message is shown.

Functionality is not implemented correctly. Steps to reproduce:
1. Start app and log in.
2. Select "BirkenBorger1".
3. Press "Tilføj ugeplan".
4. Press "Vælg billede til ugeplan".
5. Press "Tilføj fra galleri".
6. Turn on flight mode.
7. Press "Tryk for at vælge billede".
8. Select an image.
9. Enter a name.
10. Press "Gem billede".
11. **No error message is shown.**
12. Press on the newly added image.
13. Before the native file selection dialogue is shown, the error message is displayed.
14. Press back, and see that the error message is there.

The code for showing the dialogue should be around line 101 in `upload_image_from_phone_screen.dart`

**Review by giano574 — CHANGES_REQUESTED** [review]
- I get `NoSuchMethodError: The getter 'isNotEmpty' was called on null.` at `package:weekplanner/blocs/upload_from_gallery_bloc.dart:65` after picking an image (which sometimes crashes the app).
- Pressing the back button while on the "Tilføj from galleri" screen goes back to "Ny ugeplan" screen instead of "Piktogram" screen.
- There are no tests to show that the error message is shown.

Functionality is not implemented correctly. Steps to reproduce:
1. Start app and log in.
2. Select "BirkenBorger1".
3. Press "Tilføj ugeplan".
4. Press "Vælg billede til ugeplan".
5. Press "Tilføj fra galleri".
6. Turn on flight mode.
7. Press "Tryk for at vælge billede".
8. Select an image.
9. Enter a name.
10. Press "Gem billede".
11. **No error message is shown.**
12. Press on the newly added image.
13. Before the native file selection dialogue is shown, the error message is displayed.
14. Press back, and see that the error message is there.

The code for showing the dialogue should be around line 101 in `upload_image_from_phone_screen.dart`

**Review by Laexander — CHANGES_REQUESTED** [review]
When the pictogram failed to upload, the error is not displayed at the correct time, it will only be displayed if the "gem billede" button is pressed a second time.
Steps to reproduce:

1. Start app and log in.
2. Select "BirkenBorger1".
3. Press "Tilføj ugeplan".
4. Press "Vælg billede til ugeplan".
5. Press "Tilføj fra galleri".
6. Turn on flight mode.
7. Press "Tryk for at vælge billede".
8. Select an image.
9. Enter a name.
10. Press "Gem billede".
11. No error message is shown.
12. Press "Gem billede" again
13. Error message is shown

**Review by giano574 — COMMENTED** [review]
Not so good that `_checkInternetConnection` is used. If such a feature should exist, it should be a part of the API client.

Also, in this example, an error message will only be shown if there is no internet connection. If an error happens on the server, the user doesn't have the rights to upload, etc., no error will be shown.

My suggestion is something along the lines of this (on the "Gem billede" button):
```dart
onPressed: () async {
  try {
    await _uploadFromGallery.createPictogram();
  } catch(Exception e) {
    // Show error message here.
  }
}
```
This requires changing `_uploadFromGallery.createPictogram()` to return a `Future`. That is quite easy though.

Create a `Completer<PictogramModel> completer` instance as a local variable. In the existing `listen` callback, call `completer.complete(pictogram)` as the last step. In the `onError` callback, call `completer.completeError(error)` as the last step. Return `completer.future` in the last line of the method.

Alternatively, return the `Observable` from the method, change `.listen` to `.map` and change `onError:` to `.doOnError(`. Then in the `onPressed` do
```dart
onPressed: () {
  _uploadFromGallery.createPictogram()
    .listen((_) => true,
    onError: (Object error) {
      // Show error message here.
  });
}
```

File: lib/blocs/upload_from_gallery_bloc.dart [hunk 2]
```diff
@@ -95,23 +100,21 @@ class UploadFromGalleryBloc extends BlocBase {
...
         .flatMap((PictogramModel pictogram) {
       return _api.pictogram.updateImage(pictogram.id, _encodePng(_file.value));
-    }).listen((PictogramModel pictogram) {
-      // TODO(scarress):  add proper succeses handling, https://github.com/aau-giraf/weekplanner/issues/245
+    }).map((PictogramModel pictogram) {
       _isUploading.add(false);
       _pictogram.add(pictogram);
-    }, onError: (Object error) {
-      // TODO(scarress):  add error handling, https://github.com/aau-giraf/weekplanner/issues/245
+    }).doOnError((Object error) {
```
Comment (giano574) [inline lib/blocs/upload_from_gallery_bloc.dart]:
```suggestion
    }).doOnError((Object error, StackTrace trace) {
```

File: lib/blocs/upload_from_gallery_bloc.dart [hunk 3]
```diff
@@ -95,23 +100,21 @@ class UploadFromGalleryBloc extends BlocBase {
...
       accessLevel: _accessLevel,
       title: _pictogramName,
     ))
         .flatMap((PictogramModel pictogram) {
       return _api.pictogram.updateImage(pictogram.id, _encodePng(_file.value));
-    }).listen((PictogramModel pictogram) {
-      // TODO(scarress):  add proper succeses handling, https://github.com/aau-giraf/weekplanner/issues/245
+    }).map((PictogramModel pictogram) {
       _isUploading.add(false);
       _pictogram.add(pictogram);
```
Comment (giano574) [inline lib/blocs/upload_from_gallery_bloc.dart]:
```suggestion
      return pictogram;
```

File: lib/blocs/upload_from_gallery_bloc.dart [hunk 4]
```diff
@@ -40,6 +43,8 @@ class UploadFromGalleryBloc extends BlocBase {
       BehaviorSubject<String>.seeded('Offentlig');
   final BehaviorSubject<bool> _isUploading =
       BehaviorSubject<bool>.seeded(false);
+  final BehaviorSubject<bool> _hasInternet =
```
Comment (giano574) [inline lib/blocs/upload_from_gallery_bloc.dart]:
```suggestion

```

File: lib/blocs/upload_from_gallery_bloc.dart [hunk 5]
```diff
@@ -40,6 +43,8 @@ class UploadFromGalleryBloc extends BlocBase {
...
+        BehaviorSubject<bool>.seeded(true);
```
Comment (giano574) [inline lib/blocs/upload_from_gallery_bloc.dart]:
```suggestion

```

File: lib/blocs/upload_from_gallery_bloc.dart [hunk 6]
```diff
@@ -31,6 +31,9 @@ class UploadFromGalleryBloc extends BlocBase {
   /// Publishes if the input fields are filled
   Observable<bool> get isInputValid => _isInputValid.stream;
 
+  /// State of upload
+  Observable<bool> get hasInternet => _hasInternet.stream;
+
   final PublishSubject<PictogramModel> _pictogram =
```
Comment (giano574) [inline lib/blocs/upload_from_gallery_bloc.dart]:
```suggestion

```

File: lib/blocs/upload_from_gallery_bloc.dart [hunk 7]
```diff
@@ -31,6 +31,9 @@ class UploadFromGalleryBloc extends BlocBase {
   /// Publishes if the input fields are filled
   Observable<bool> get isInputValid => _isInputValid.stream;
 
+  /// State of upload
+  Observable<bool> get hasInternet => _hasInternet.stream;
```
Comment (giano574) [inline lib/blocs/upload_from_gallery_bloc.dart]:
```suggestion

```

File: lib/blocs/upload_from_gallery_bloc.dart [hunk 8]
```diff
@@ -31,6 +31,9 @@ class UploadFromGalleryBloc extends BlocBase {
   /// Publishes if the input fields are filled
   Observable<bool> get isInputValid => _isInputValid.stream;
 
+  /// State of upload
```
Comment (giano574) [inline lib/blocs/upload_from_gallery_bloc.dart]:
```suggestion

```

**Review by Laexander — DISMISSED** [review]

File: lib/screens/upload_image_from_phone_screen.dart [hunk 9]
```diff
@@ -130,13 +138,27 @@ class UploadImageFromPhone extends StatelessWidget {
...
     );
   }
 
+  void _showUploadError(BuildContext context) {
+    showDialog<Center>(
+        context: context,
+        barrierDismissible: false,
+        builder: (BuildContext context) {
+          return const GirafNotifyDialog(
+            key: Key('UploadErrorKey'),
```
Comment (giano574) [inline lib/screens/upload_image_from_phone_screen.dart]:
```suggestion

```
Not used.

File: test/screens/upload_image_from_phone_screen_test.dart [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    di.clearAll();
+    di.registerDependency<UploadFromGalleryBloc>((_) => bloc);
+    di.registerDependency<ToolbarBloc>((_) => ToolbarBloc());
+    di.registerDependency<AuthBloc>((_) => AuthBloc(api));
+  });
+
+  testWidgets('Tests error dialog pops up on upload error',
+      (WidgetTester tester) async {
+    await tester.pumpWidget(MaterialApp(
+        home: UploadImageFromPhone()
```
Comment (giano574) [inline test/screens/upload_image_from_phone_screen_test.dart]:
```suggestion
        home: UploadImageFromPhone(),
```

File: test/screens/upload_image_from_phone_screen_test.dart [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    ));
+    await tester.pumpAndSettle();
+    when(api.pictogram.create(any))
+        .thenAnswer((_) => Observable<PictogramModel>.error(Exception()));
+    bloc.setInputIsValid(true);
+
+    await tester.tap(find.byKey(const Key('SavePictogramButtonKey')));
+    await tester.pumpAndSettle();
+    expect(find.text('Fejl'), findsOneWidget);
```
Comment (giano574) [inline test/screens/upload_image_from_phone_screen_test.dart]:
```suggestion
    expect(find.text('Upload af pictogram fejlede.'), findsOneWidget);
```

File: lib/screens/upload_image_from_phone_screen.dart [hunk 12]
```diff
@@ -130,13 +138,27 @@ class UploadImageFromPhone extends StatelessWidget {
...
+            title: 'Fejl',
+            description: 'Upload af pictogram fejlede.',
+          );
+        }
```
Comment (giano574) [inline lib/screens/upload_image_from_phone_screen.dart]:
```suggestion
        },
```

**Review by giano574 — COMMENTED** [review]
Probably autoformat all the files also

**Review by giano574 — APPROVED** [review]

**Review by Laexander — APPROVED** [review]

**Pull request merged**